### PR TITLE
Fix propagation order

### LINKED ISSUE #949: A better abstraction than `propagateToChildren`
# What could be improved
Code that is using `propagateToChildren` repeats some snippets of code a lot. 
It might be possible to create a better abstraction than `propagateToChildren` that requires each place to have less code...

# Why should this be improved
Less code duplication.

# Any risks?
It could be less efficient than `propagateToChildren`, very unlikely though.

# More information

### LINKED ISSUE #1175: Gesture dispatch error  when in overlapped tappables.
Demo code:

```
class TestGame extends FlameGame with HasTappables {
  @override
  Future<void> onLoad() {
    add(ParentComponent());
    return super.onLoad();
  }
}

class ParentComponent extends PositionComponent {
  @override
  Future<void> onLoad() {
    add(ChildComponent('child1'));
    add(ChildComponent('child2'));
    return super.onLoad();
  }
}

class ChildComponent extends PositionComponent with Tappable {
  final String value;

  ChildComponent(this.value);

  @override
  Future<void> onLoad() {
    size = Vector2.all(50);
    return super.onLoad();
  }

  @override
  bool onTapUp(TapUpInfo info) {
    print(value);
    return false;
  }
}

```

Expect `child2` but get `child1`.

I think it's caused by the for loop in `Component.dart#propagateToChildren()`

### PR DESCRIPTION
**Pull request by spydon**
# Description

Since the children weren't reversed in `propagateToChildren` the events were not first passed to the last added component.
In `OrderedSet` we now cache the reversed set, so it is a slightly less impact than it should have been to have reversed here than before.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors and are passing (See [Contributor Guide]).
- [x] My PR does not decrease the code coverage, or I have __a very special case__ and explained on the PR description why this PR decreases the coverage.
- [x] I updated/added relevant documentation (doc comments with `///`) and updated/added examples in `doc/examples`.
- [x] I have formatted my code with `flutter format` and the `flutter analyze` does not report any problems.
- [x] I read and followed the [Flame Style Guide].
- [x] I have added a description of the change under `[next]` in `CHANGELOG.md`.
- [x] I removed the `Draft` status, by clicking on the `Ready for review` button in this PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flame users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in `CHANGELOG.md`).
- [x] No, this is *not* a breaking change.

## Related Issues

Fixes #1175
Closes #949 

[issue database]: https://github.com/flame-engine/flame/issues
[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Flame Style Guide]: https://github.com/flame-engine/flame/blob/main/STYLEGUIDE.md

### CODE REVIEW

File: packages/flame/test/components/composed_component_test.dart [hunk 1]
```diff
@@ -148,6 +148,32 @@ void main() {
...
+        await parent.ensureAdd(child2);
+        game.onTapDown(
+          1,
+          createTapDownEvent(
+            game,
+            globalPosition: const Offset(50, 50),
+          ),
+        );
+
+        expect(child2.tapped, true);
```
Comment (erickzanardo) [inline packages/flame/test/components/composed_component_test.dart:170]:
```suggestion
        expect(child2.tapped, isTrue);
```

File: packages/flame/test/components/composed_component_test.dart [hunk 2]
```diff
@@ -148,6 +148,32 @@ void main() {
...
+        expect(child2.tapTimes, 1);
```
Comment (erickzanardo) [inline packages/flame/test/components/composed_component_test.dart:171]:
```suggestion
        expect(child2.tapTimes, equals(1));
```

File: packages/flame/test/components/composed_component_test.dart [hunk 3]
```diff
@@ -148,6 +148,32 @@ void main() {
...
+        expect(child1.tapped, false);
```
Comment (erickzanardo) [inline packages/flame/test/components/composed_component_test.dart:172]:
```suggestion
        expect(child1.tapped, isFalse);
```

File: packages/flame/test/components/composed_component_test.dart [hunk 4]
```diff
@@ -148,6 +148,32 @@ void main() {
...
+        expect(child1.tapTimes, 0);
```
Comment (erickzanardo) [inline packages/flame/test/components/composed_component_test.dart:173]:
```suggestion
        expect(child1.tapTimes, equals(0));
```

**Review by erickzanardo — APPROVED** [review]
Some suggestions, but LGTM already

**Review by luanpotter — APPROVED** [review]
YES! YES! YES!

**Pull request merged**